src/multiclient/multi_client_actor.cpp

```diff
@@ -0,0 +1,266 @@
+    worker.check_retry_count = 0;
+  } else {
+    worker.check_retry_after = td::Timestamp::in(kRetryInterval);
+  }
+}
+
+void MultiClientActor::check_archival() {
+  static constexpr int32_t kBlockWorkchain = ton::masterchainId;
+  static constexpr int64_t kBlockShard = ton::shardIdAll;
+  static constexpr int32_t kBlockSeqno = 3;
+
+  static constexpr int kLookupMode = 1;
+  static constexpr int kLookupLt = 0;
+  static constexpr int kLookupUtime = 0;
+
+  for (size_t worker_index = 0; worker_index < workers_.size(); worker_index++) {
+    if (!workers_[worker_index].is_alive) {
+      continue;
+    }
+
+    send_worker_request<ton::tonlib_api::blocks_lookupBlock>(
+        worker_index,
+        ton::tonlib_api::blocks_lookupBlock(
+            kLookupMode,
+            ton::tonlib_api::make_object<ton::tonlib_api::ton_blockId>(kBlockWorkchain, kBlockShard, kBlockSeqno),
+            kLookupLt,
+            kLookupUtime
+        ),
+        [self_id = actor_id(this), worker_index](auto result) {
+          td::actor::send_closure(self_id, &MultiClientActor::on_archival_checked, worker_index, result.is_ok());
+        }
+    );
+  }
+}
+
+void MultiClientActor::on_archival_checked(size_t worker_index, bool is_archival) {
+  LOG(DEBUG) << "LS #" << worker_index << " archival: " << is_archival;
+  workers_[worker_index].is_archival = is_archival;
+}
+
+std::vector<size_t> MultiClientActor::select_workers(const RequestParameters& options) const {
+  std::vector<size_t> result;
+  result.reserve(workers_.size());
+
+  for (size_t i : std::views::iota(0u, workers_.size()) |
+           std::views::filter([&](size_t i) { return workers_[i].is_alive; }) |
+           std::views::filter([&](size_t i) { return options.archival == true ? workers_[i].is_archival : true; })) {
+    result.push_back(i);
+  }
+
+  if (result.empty()) {
+    return result;
+  }
+
+  switch (options.mode) {
+    case RequestMode::Broadcast:
+      return result;
+
+    case RequestMode::Single: {
+      if (options.lite_server_indexes.has_value()) {
+        CHECK(!options.lite_server_indexes->empty() && options.lite_server_indexes->size() == 1);
+
+        return std::find(result.begin(), result.end(), options.lite_server_indexes->front()) != result.end() ?
+            std::vector<size_t>{options.lite_server_indexes.value().front()} :
+            std::vector<size_t>{};
+      }
+
+      return std::vector<size_t>{result[get_random_index<size_t>(0, result.size() - 1)]};
+    }
+
+    case RequestMode::Multiple: {
+      CHECK(!(options.clients_number.has_value() && options.lite_server_indexes.has_value()));
+      CHECK(options.clients_number.has_value() || options.lite_server_indexes.has_value());
+
+      if (options.lite_server_indexes.has_value()) {
+        std::set<size_t> result_set(result.begin(), result.end());
+        for (auto index : options.lite_server_indexes.value()) {
+          if (!result_set.contains(index)) {
+            result_set.erase(index);
```

Comment:
should it be `insert` ?

Reply:
Not sure I get it, `options.lite_server_indexes` means ls indexes that we want to filter out from selected workers? In that case there should be no negation in `if` statement.

Reply:
it has to work like that:
- we have 4 workers [1, 2, 3, 4]
- the first 3 of them are alive and fit to requirements
- we want to send request only to 1st and 4th: [1, 4]
so initially we have filtered list [1, 2, 3] and then we check which one contains in the preferred list of [1, 4]. If it doesn't — we remove it.

Reply:
Got it, so we need intersection of `result` and `options.lite_server_indexes`. But the code will return [1,2,3] in your example.

Reply:
yeah. sry, my bad. fixed in the next commit